torch/utils/checkpoint.py

```diff
@@ -1587,6 +1592,40 @@ def recompute_fn(*inputs):
 
     return
 
+
+class GraphExecGroup:
+    """Any checkpointed regions encountered by backward under the same instance
+    of this context manager will only trigger recompute at most once, even if
+    there are multiple calls to backward.
+
+    Backward calls under the same instance of this context manager must execute
+    over non-overlapping regions of the backward graph even if retain_graph=True.
+    """
+    _current_group: Optional[int] = None
+    _counter: int = 0
+
+    def __init__(self) -> None:
```

Comment:
Since we never expect this to be set globally, you most likely can make this a regular context manager with something like https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager that will make your life a lot easier